Mage/src/main/java/mage/watchers/Watcher.java

```
@@ -114,96 +112,7 @@ public <T extends Watcher> T copy() {
-                            Map<Object, Set<Object>> target = (Map<Object, Set<Object>>) field.get(watcher);
-                            target.clear();
-                            for (Map.Entry<Object, Set<Object>> e : source.entrySet()) {
-                                Set<Object> set = new TreeSet<>();
-                                set.addAll(e.getValue());
-                                target.put(e.getKey(), set);
-                            }
-                        } else if (valueType.getTypeName().contains("Set")) {
-                            // Map<UUID, Set<Object>>
-                            Map<Object, Set<Object>> source = (Map<Object, Set<Object>>) field.get(this);
-                            Map<Object, Set<Object>> target = (Map<Object, Set<Object>>) field.get(watcher);
-                            target.clear();
-                            for (Map.Entry<Object, Set<Object>> e : source.entrySet()) {
-                                Set<Object> set = new HashSet<>();
-                                set.addAll(e.getValue());
-                                target.put(e.getKey(), set);
-                            }
-                        } else if (valueType.getTypeName().contains("PlayerList")) {
-                            // Map<UUID, PlayerList>
-                            Map<Object, PlayerList> source = (Map<Object, PlayerList>) field.get(this);
-                            Map<Object, PlayerList> target = (Map<Object, PlayerList>) field.get(watcher);
-                            target.clear();
-                            for (Map.Entry<Object, PlayerList> e : source.entrySet()) {
-                                PlayerList list = e.getValue().copy();
-                                target.put(e.getKey(), list);
-                            }
-                        } else if (valueType.getTypeName().endsWith("Cards")) {
-                            // Map<UUID, Cards>
```

Same check - make sure it copy corrected (some xmage object inherited from list or hashmap, but contains additional data and must be process as copyable).

---

I added `Copyable<Cards>` to the Cards interface, so that's handled for that case specifically. Is there a way to check for this situation in general?

---

Its interesting. Maybe copy code must have additional check for data classes to catch. You can find instances but not extended classes by: https://stackoverflow.com/a/16688794 : getClass().equals(Foo.class)

I don't know is it a problem at all.

---

Check order example:
- immutable
- copyable
- known classes
- default classes (same name like ArrayList, HashMap, etc);
- non default classes (instanceof Map, Hash, etc);

The last one can catch bad xmage objects with miss copyable code in theory.

---

I think that it's better to throw an exception than to silently change to the wrong type of class. That was a good suggestion, turns out there were also `LinkedHashMap` and `LinkedHashSet` which should be handled separately.